python/ray/ray_operator/operator.py

```diff
@@ -181,162 +169,108 @@ def write_config(self) -> None:
-        self._generation = generation
-
-    def set_num_retries(self, num_retries: int) -> None:
-        self._num_retries = num_retries
-
-    def get_generation(self) -> int:
-        return self._generation
-
-    def get_num_retries(self) -> int:
-        return self._num_retries
+        try:
+            os.remove(self.config_path)
+        except OSError:
+            log_prefix = ",".join([self.name, self.namespace])
+            logger.warning(f"{log_prefix}: config path does not exist {self.config_path}")
 
 
 # Maps ray cluster (name, namespace) pairs to RayCluster python objects.
 ray_clusters = {}  # type: Dict[Tuple[str, str], RayCluster]
 
 
-def run_event_loop():
-    # Instantiate event stream.
-    if operator_utils.NAMESPACED_OPERATOR:
-        raycluster_cr_stream = operator_utils.namespaced_cr_stream(
-            namespace=operator_utils.OPERATOR_NAMESPACE)
-    else:
-        raycluster_cr_stream = operator_utils.cluster_scoped_cr_stream()
-
-    # Run control loop.
-    for event in raycluster_cr_stream:
-        cluster_cr = event["object"]
-        cluster_name = cluster_cr["metadata"]["name"]
-        cluster_namespace = cluster_cr["metadata"]["namespace"]
-        event_type = event["type"]
-        handle_event(event_type, cluster_cr, cluster_name, cluster_namespace)
-
-
-def handle_event(event_type, cluster_cr, cluster_name, cluster_namespace):
-    # TODO: This only detects errors in the parent process and thus doesn't
-    # catch cluster-specific autoscaling failures. Fix that (perhaps at
-    # the same time that we eliminate subprocesses).
-    try:
-        cluster_action(event_type, cluster_cr, cluster_name, cluster_namespace)
-    except Exception:
-        log_prefix = ",".join(cluster_name, cluster_namespace)
-        if event_type in ["ADDED", "MODIFIED"]:
-            logger.exception(f"{log_prefix}: Error while updating RayCluster.")
-            cluster_status_q.put((cluster_name, cluster_namespace,
-                                  STATUS_ERROR))
-        elif event_type == "DELETED":
-            # Don't try to update CRD's status if the CRD is gone.
-            logger.exception(
-                f"Error while deleting RayCluster {cluster_name}.")
-
-
-def cluster_action(event_type: str, cluster_cr: Dict[str, Any],
-                   cluster_name: str, cluster_namespace: str) -> None:
-
-    cluster_config = operator_utils.cr_to_config(cluster_cr)
-    cluster_identifier = (cluster_name, cluster_namespace)
-    log_prefix = ",".join(cluster_identifier)
+@kopf.daemon('rayclusters')
+def status_handling_loop(stopped, name, namespace, **kwargs):
+    cluster_identifier = (name, namespace)
+    while not stopped:
+        time.sleep(1)
+
+        ray_cluster = ray_clusters.get(cluster_identifier)
+        if not ray_cluster:
+            continue
+
+        cluster_status_q = ray_cluster.cluster_status_q
+        if cluster_status_q.empty():
+            continue
```

Comment:
Nice, definitely makes sense to have a queue per RayCluster.

What are ramifications of watching for at most one status update per second with a non-blocking queue vs. watching for all updates with a blocking queue.

Reply:
The sleep / non-blocking call is just to allow the `stop` signal to be picked up in a timely manner to prevent .

> If I understand right, a watch on an k8s api object just responds to updates to the object's `resourceVersion`.
> I imagine @kopf.on.update probably does more or less the same thing under the hood? Maybe with some sort of improvement to robustness? Curious to take a look and see...
> 
> I don't think the issues observed with the Ray 1.3.0 are with the watch itself -- it's mostly that events are handled plain wrong.
> That said, looks like using kopf could definitely make things simpler and less error-prone.

The biggest difference between using the watch and `on.update` is around robustness, retries, etc. See this section for a more detailed overview of some feature kopf provides in this area: https://kopf.readthedocs.io/en/stable/continuity/

This is not something that watching the event stream alone would provide.

Reply:
Both approaches I think are not ideal. Even with this approach, we run the risk of dropping status updates if a pod failure occurs before the update is picked up by the consumer here. A more stable solution would be to set an annotation on the ray cluster object (instead of enquring a status update) and then registering a handler for these changes:

https://kopf.readthedocs.io/en/stable/handlers/#field-handlers

This way, even if a pod failure occurs, we'll pick up the status update and be able to apply it correctly. What do you think?

Reply:
As a short term solution, because this change might require some more thought, I found a way of setting up and tearing down the threads that more closely mirrors the original implementation. Let me know what you think.

Reply:
Current solution looks good for this PR.
I agree that in a future PR, Status updates should be handled more robustly using Kopf primitives.